### ISSUE TITLE
(V3 Migration) okhttp.OkHttpClient: <-- HTTP FAILED: java.lang.IllegalStateException: closed 

### ISSUE THREAD

**Issue by kodekid20** [apollographql/apollo-kotlin#4054]
Hi, I am migrating to V3, but keep getting the following error when uploading file(s).
>okhttp.OkHttpClient: <-- HTTP FAILED: java.lang.IllegalStateException: closed 

In version 2, everything works fine based on my implementation below: 
```kotlin
apolloClient.mutate(
     CreateProjectMutation(
          ...
          video = FileUpload(mimetype = type, filePath = parameters.video),
          thumbnail = FileUpload(mimetype = thumbType, filePath = parameters.thumbnail)
       )
 )
```

This is my implementation in V3:
```kotlin
val upload = DefaultUpload.Builder()
      .content(File(parameters.video)) // using DefaultUpload.Builder content ext
      .contentType(type)
      .build()

val thumbnail = DefaultUpload.Builder()
      .content(File(parameters.thumbnail))
      .contentType(thumbType)
      .build()

apolloClient.mutation(
     CreateProjectMutation(
          ...
          video = upload,
          thumbnail = thumbnail
       )
 )
```

For reference, here is my mutation
```graphql
mutation CreateProject(
    ...
    $thumbnail: Upload!,
    $video: Upload!
) ...
```

Here's the network log
```okhttp
I/okhttp.OkHttpClient: Content-Type: multipart/form-data; boundary=44a3c556-0e96-44e1-824a-f3bfa67322a3
I/okhttp.OkHttpClient: X-APOLLO-OPERATION-ID: 3ad17cbe8d5f05812ba6d19dabd9f896304b6cb1808d105b0697ee188a074dc9
I/okhttp.OkHttpClient: X-APOLLO-OPERATION-NAME: CreateProject
I/okhttp.OkHttpClient: Authorization: JWT eyJ0eXAiOiJKV1QiLCJhb...
I/okhttp.OkHttpClient: apollographql-client-name: Android
I/okhttp.OkHttpClient: apollographql-client-version: 1.0.0
...
I/okhttp.OkHttpClient: �e��T��o�~����VJ/6�y{�_a�Op����/���5���.����˂�Η���˷)�i�Vj�"�Ô��3q�2a�J
I/okhttp.OkHttpClient: ZZZZZZZ^
I/okhttp.OkHttpClient: --> END POST (-1-byte body)
I/okhttp.OkHttpClient: <-- HTTP FAILED: java.lang.IllegalStateException: closed
```

### What I did
I checked to make sure nothing was calling `response.body.toString()` 2x. Double checked the image/video files that were passed to `DefaultUpload.Builder`.  Everything seems to be in place, but I am now going nuts trying to find the issue. Any idea of what I am doing wrong or missing? Any help would be greatly appreciated. Thanks

**Comment by BoD** [comment]
Hi 👋  Thanks for reporting this.

Hmm the bits of code you're showing seem correct to me. Would it be possible for you to inspect the payloads with something like [Charles Proxy](https://www.charlesproxy.com/)? With that you could compare the exchanges with v2 and v3.

**Comment by kodekid20** [comment]
Thank you for responding! I will inspect the payloads for both and report back.

**Comment by kodekid20** [comment]
@BoD After inspecting the payloads, here are the results:

From V2
```log
:method: POST
:path: /graphql/
:authority: <domain>
:scheme: https
accept: application/json
x-apollo-operation-id: c79696dd0d287586fb94e3c99c380e13d9ae4638b5061212a443f431a9cb0857
x-apollo-operation-name: CreateProject
authorization: JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJ...
apollographql-client-name: Android
apollographql-client-version: 1.0.0
content-type: multipart/form-data; boundary=f0e40b0d-dc6f-4dbb-a71c-f865a37c7e54
content-length: 22491137
accept-encoding: gzip
user-agent: okhttp/4.9.3

--c0cf8d1f-41ff-4f8c-894f-42995fdbbd69
Content-Disposition: form-data; name="operations"
Content-Type: application/json; charset=utf-8
Content-Length: 2050

{"operationName":"CreateProject",... }"}
--c0cf8d1f-41ff-4f8c-894f-42995fdbbd69
Content-Disposition: form-data; name="map"
Content-Type: application/json; charset=utf-8
Content-Length: 53

{"0":["variables.thumbnail"],"1":["variables.video"]}
--c0cf8d1f-41ff-4f8c-894f-42995fdbbd69
Content-Disposition: form-data; name="0"; filename="1650494494261.png"
Content-Type: image/png
Content-Length: 2803375

PNG
...

--c0cf8d1f-41ff-4f8c-894f-42995fdbbd69
Content-Disposition: form-data; name="1"; filename="1650494481410_20220205_152829.mp4"
Content-Type: video/mp4
Content-Length: 5492006
...
```

```log
Multipart

Part                       Content-type                                     Filename                                                               Value
operations	      application/json; charset=utf-8		                                                                        2.01 KB (2,058 bytes)
map           	      application/json; charset=utf-8		                                                                        {"0":["variables.thumbnail"],"1":["variables.video"]}
0	                      image/png	                                 1650493693683.png	                                        2.21 MB (2,316,430 bytes)
1	                      video/mp4	                                 1650493658148_20220405_173552.mp4	          13.79 MB (14,458,009 bytes)
```

V3 fails out the gate
```log
:method	POST
:path	/graphql/
:authority	<domain>
:scheme	https
x-apollo-operation-id	3ad17cbe8d5f05812ba6d19dabd9f896304b6cb1808d105b0697ee188a074dc9
x-apollo-operation-name	CreateProject
authorization	JWT eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.ey...
apollographql-client-name	Android
apollographql-client-version	1.0.0
content-type	multipart/form-data; boundary=f3147170-9411-4b8f-bb91-3b89eed09db6
accept-encoding	gzip
user-agent	okhttp/4.9.3
```
From Charles Proxy: `Failure | IO: Stream cancelled by CLIENT`
IDE console: `I/okhttp.OkHttpClient: <-- HTTP FAILED: java.lang.IllegalStateException: closed`

**Comment by kodekid20** [comment]
@BoD One thing I notice in DefaultHttpRequestComposer.kt is that if `uploads` Is not empty `contentLength` is set to -1L

```kotlin
// XXX: support non-chunked multipart
override val contentLength = -1L
```

That is the only difference I see right now.

**Comment by BoD** [comment]
Thanks a lot for this!

When setting the `contentLength` to -1, OkHttp normally sets a `Transfer-Encoding` header to `chunked` ([source](https://github.com/square/okhttp/blob/okhttp_4.9.x/okhttp/src/main/kotlin/okhttp3/internal/http/BridgeInterceptor.kt#L53)) but I don't see this header in your trace, this is odd 🤔 

I guess it is possible that the server doesn't support chunked and requires a `Content-Length` but that still wouldn't explain why we don't see the header.

Would be possible for us to access the server so we can investigate on our side? (If it is not public / requires authentication, you can send information privately to my email benoit.lubek@apollographql.com).

**Comment by BoD** [comment]
Hi again @kodekid20, thanks for sending the info, that helped a lot!

Indeed it looks like the culprit is the lack of `Content-Length` header, as it started working as soon as I added it.

I've made a project [here](https://github.com/BoD/apollo-kotlin-repro-issue-4054/blob/master/src/main/kotlin/Main.kt#L25) which shows a workaround that adds the `Content-Length` header with an `HttpInterceptor` (thanks @martinbonnin for the idea!). The problem is that in order to do this, it reads the whole payload into memory, which could lead to OutOfMemoryErrors. (Maybe there's a way to know the length without reading it in memory but we're not sure how for now)  But it may be a temporary solution to unblock you until we have a real fix.

(Still not sure why we can't see the `Transfer-Encoding` header but I think it's just Charles Proxy not showing everything).

**Comment by kodekid20** [comment]
Thank you @BoD  and @martinbonnin for the assist! I will keep an eye out for the fix. I will adopt the suggestion in the meantime.

**Comment by BoD** [comment]
The fix has been merged so it will be in the next release, in the meantime it can be tried out with the [snapshots](https://github.com/apollographql/apollo-kotlin#snapshots).

**Comment by BoD** [comment]
Available in v3.3.0.

**Comment by BoD** [comment]
Hi @sterien7! 

If you are using 3.3.0, the fix to send the `Content-Length` header is there, so you may be seeing another issue.

Can you tell us a bit more about what you're doing and the issue you're seeing?

**Comment by sterien7** [comment]
@BoD I am debugging the app and I think that the `HttpLoggingInterceptor` is causing the issue.
So the buffer is closed because the interceptor has printed it. After I removed it the HTTP request fails and I get `Failed to execute GraphQL http network request`

**Comment by sterien7** [comment]
@BoD I found the solution which is to add the filename so the builder looks like this
```
DefaultUpload
        .Builder()
        .content(file)
        .contentType(it.mimeType)
        .fileName(it.mediaUrl)
        .build()
```

The issue it I cannot run the logger to see what is going on with the calls

**Comment by BoD** [comment]
@sterien7 Good to know you found out the culprit! And thanks for letting us know of the solution.

About the logger: you are right, the way it works is that it consumes the bytes of the uploads in order to print the body, which will make the rest of the chain fail later. Maybe there is a way to improve this (or at least warn in the documentation!) I'll have a look.

**Comment by sterien7** [comment]
@BoD yeah the logger is `okhttp3.logging.HttpLoggingInterceptor` and if you set it up with `HttpLoggingInterceptor.Level.BASIC` which is not logging the http body it works but if you change the level to `BODY` then it consumes the bytes so they are not there to upload them afterwards. Keep in mind that with the old Apollo library I used till yesterday `1.4.4` I haven't had that issue, if that helps somehow

**Comment by BoD** [comment]
Thanks a lot! In fact I didn't realize you were using the okhttp interceptor, I thought you were using the Apollo one (`com.apollographql.apollo3.network.http.LoggingInterceptor` which has the same issue). This is good information 🙏

**Comment by BoD** [comment]
Just a note to say that with #4125, using OkHttp's (or Apollo's) logging interceptor should no longer break uploading files. This will be available in the next release.

**Comment by damianpetla-bright** [comment]
There is one more case where this fail and it will be worth testing. I am using Apollo in Android app so I use Android Studio. It has _App Inspection_ feature that automatically allow monitoring network operations if OkHttp is used. Whenever that inspector is active, uploading files fails with `java.lang.IllegalStateException: closed`

**Comment by martinbonnin** [comment]
@damianpetla-bright can you try with [the SNAPSHOTs](https://github.com/apollographql/apollo-kotlin#snapshots)? @BoD made a couple of changes there recently that could help with this.

**Comment by damianpetla-bright** [comment]
@martinbonnin tried version `3.3.2-SNAPSHOT` but issue is still there. Below is the exception thrown. It's easy to reproduce. Just try uploading file while App Inspection in Android Studio is active. Fun fact is that I don't really need Network monitor from App Inspector but it comes with Work manager inspector an DB inspector that I cannot use now which makes development much harder. In my own interceptors I buffering body as Apollo library expect but I cannot change interceptor used by IDE :( 
Hope that you can help with that somehow.

```kotlin
 Could not track an OkHttp3 request
    java.lang.IllegalStateException: Apollo: DefaultUpload body can only be read once. If you want to read it several times for logging or other purposes, either buffer it in memory or use your own `Upload` implementation.
        at com.apollographql.apollo3.api.DefaultUpload.writeTo(DefaultUpload.kt:27)
        at com.apollographql.apollo3.api.http.UploadsHttpBody.writeBoundaries(DefaultHttpRequestComposer.kt:319)
        at com.apollographql.apollo3.api.http.UploadsHttpBody.writeTo(DefaultHttpRequestComposer.kt:278)
        at com.apollographql.apollo3.network.http.DefaultHttpEngine$execute$2$httpRequest$1$2.writeTo(OkHttpEngine.kt:60)
        at com.android.tools.profiler.agent.okhttp.OkHttp3Interceptor.trackRequest(OkHttp3Interceptor.java:91)
        at com.android.tools.profiler.agent.okhttp.OkHttp3Interceptor.intercept(OkHttp3Interceptor.java:45)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at com.android.tools.appinspection.network.okhttp.OkHttp3Interceptor.intercept(OkHttp3Interceptor.kt:46)
```

**Comment by damianpetla-bright** [comment]
I have resolved the issue for me with using `content(ByteArray)` instead of `content(File)` since files I am uploading aren't that big (few MB).

**Comment by BoD** [comment]
Thanks for the update! 

I can confirm I'm reproducing the issue.

The problem comes from the OkHttp Interceptor that Android Studio injects for the inspector, which reads (and consumes) the `Upload`'s body once before forwarding down the chain, where it's expectedly read again.

I can see that this has been causing problems to other people in the past, and an issue was opened on Android Studio's issue tracker, but was closed without a resolution. I've opened a new one [here](https://issuetracker.google.com/issues/235863810).

I could think of a (hacky!) workaround, which is to detect the presence of the inspector and deny the body reading:

```kotlin
class UploadWrapper(private val upload: Upload) : Upload {
    override val contentLength: Long
        get() = upload.contentLength

    override val contentType: String
        get() = upload.contentType

    override val fileName: String?
        get() = upload.fileName

    override fun writeTo(sink: BufferedSink) {
        val stacktrace = Throwable().stackTrace
        val isFromNetworkInspector = stacktrace.any { it.className.startsWith("com.android") && it.methodName == "trackRequest" }
        if (isFromNetworkInspector) {
            // Do not consume the Upload when from Android Studio's Network Inspector
        } else {
            upload.writeTo(sink)
        }
    }
}
```

It can be used like so:

```kotlin
val upload = DefaultUpload.Builder()
  // ...
  .build()

apolloClient.mutation(UploadMutation(UploadWrapper(upload))).execute()
```

> I have resolved the issue for me using content(ByteArray) instead of content(File) since files I am uploading aren't that big (few MB).

If your files are not too big indeed this is a solution, but it is not ideal as they are loaded into memory, which could get full. An idea: you could have this behavior (or the hack above) only in debug builds.

**Comment by martinbonnin** [comment]
https://github.com/apollographql/apollo-kotlin/pull/4228 added the possibility to reuse a `File` upload. You can now use `File.toUpload()` for this. It'll be available in the SNAPSHOT in ~40min. Let us know how that goes!

**Comment by damianpetla-bright** [comment]
@BoD Used your solution and works as planned. Additionally, used Kotlin delegation to make wrapper even shorter
```kotlin
class UploadWrapper(private val upload: Upload) : Upload by upload {

    override fun writeTo(sink: BufferedSink) {
        val stacktrace = Throwable().stackTrace
        val isFromNetworkInspector =
            stacktrace.any { it.className.startsWith("com.android") && it.methodName == "trackRequest" }
        if (isFromNetworkInspector) {
            //Do not consume the Upload when from Android Studio's Network Inspector
        } else {
            upload.writeTo(sink)
        }
    }
}
```

**Comment by martinbonnin** [comment]
Re-closing this one as 3.4.0+ `File.toUpload()` now allows reading the file multiple times. Let us know if this is happening again.
